vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,615 @@
+            let page = self.pages_before_region + page;
+            self.logmem[page_word(page)].fetch_or(1 << page_bit(page), Ordering::Relaxed);
+        }
+    }
+
+    // Check whether the specified offset is marked as dirty. The `offset` is relative
+    // to the memory region, so a `0` offset references the start of the memory region.
+    // Any attempt to access beyond the end of the bitmap are simply ignored.
+    fn dirty_at(&self, offset: usize) -> bool {
+        let page = page_number(offset);
+        if page >= self.number_of_pages {
+            return false; // ignore out of bound access
+        }
+
+        // get the absolute page number
+        let page = self.pages_before_region + page;
+        let page_bit = self.logmem[page_word(page)].load(Ordering::Relaxed) & (1 << page_bit(page));
+        page_bit != 0
+    }
+}
+
+/// `MmaplogReg` mmaps the frontend bitmap backing memory in the current process.
+#[derive(Debug)]
+pub struct MmapLogReg {
+    addr: *const AtomicU8,
+    len: usize,
+}
+
+// SAFETY: Send is not automatically implemented because the raw pointer.
+// No one besides `MmapLogReg` has the raw pointer, so we can safely transfer it to another thread.
+unsafe impl Send for MmapLogReg {}
+
+// SAFETY: Sync is not automatically implemented because the raw pointer.
+// `MmapLogReg` doesn't have any interior mutability and all access to `&AtomicU8`
+// are done through atomic operations.
+unsafe impl Sync for MmapLogReg {}
+
+impl MmapLogReg {
+    #![allow(dead_code)]
+    // Note: We could try to adjust the mapping area to only cover the memory region, but
+    // the region's starting address is not guarantee to be LOG_WORD_SIZE-page aligned
+    // which makes the implementation needlessly cumbersome.
+    // Note: The specification does not define whether the offset must be page-aligned or not.
+    // But, since we are receiving the offset from the frontend to be used to call mmap,
+    // we assume it is properly aligned (currently, qemu always send a 0 offset).
+    pub(crate) fn from_file(fd: BorrowedFd, offset: u64, len: u64) -> io::Result<Self> {
+        let offset: isize = offset.io_try_into()?;
+        let len: usize = len.io_try_into()?;
+
+        // Let's uphold the safety contract for `std::ptr::offset()`.
+        if len > isize::MAX as usize {
+            return Err(io::Error::from(io::ErrorKind::InvalidData));
+        }
+
+        // SAFETY: `fd` is a valid file descriptor and we are not using `libc::MAP_FIXED`.
+        let addr = unsafe {
+            libc::mmap(
+                ptr::null_mut(),
+                len as libc::size_t,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                fd.as_raw_fd(),
+                offset as libc::off_t,
+            )
+        };
+
+        if addr == libc::MAP_FAILED {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(Self {
+            addr: addr as *const AtomicU8,
```

Comment:
I did that in one of my earlier experiments, IIRC it doesn't make things easier, at least with the current public API it will expose the lifetime parameter. 

The `Derf` could work (not a fan, since this is not a smart-pointer like type). The long term plan is to extend use the `vm-memory  `AtomicBitmap` as  `InnerBitmap`, so it should support as log backing memory both `MmapLogReg`and `Vec`.